### PR TITLE
add tk_to_qujax_args and qujax_to_qk

### PR DESCRIPTION
**Pull request by SamDuffield**

### CODE REVIEW

File: README.md
Comment (cqc-melf) [inline README.md:19]: Is this change intended?

Comment (SamDuffield) [reply README.md]: I meant to change it to the github link (I think it's better for github to link to github)
but mistakenly linked it to pytket-qujax rather than qujax lol

File: pytket/extensions/qujax/_metadata.py
Comment (cqc-melf) [inline pytket/extensions/qujax/_metadata.py:1]: This file should no be in the git, can you remove it?

Comment (SamDuffield) [reply pytket/extensions/qujax/_metadata.py:1]: should add to .gitignore or update my local git settings?

Comment (cqc-melf) [reply pytket/extensions/qujax/_metadata.py:1]: It should have been in the gitignore already, if not please add it.

File: .gitignore
Comment (cqc-melf) [inline .gitignore]: I think this is not the right way to add this, because we want to keep the _metadata.py in the main directory

Comment (SamDuffield) [reply .gitignore]: ok! maybe better now?

File: pytket/extensions/qujax/qujax_convert.py
Comment (cqc-melf) [inline pytket/extensions/qujax/qujax_convert.py:215]: Why is this a function in pytket-qujax? What is the benefit from having this in pytket-qujax and not only in qujax?

File: pytket/extensions/qujax/qujax_convert.py
Comment (cqc-melf) [inline pytket/extensions/qujax/qujax_convert.py:183]: Why is this a function in pytket-qujax? What is the benefit from having this in pytket-qujax and not only in qujax?

File: pytket/extensions/qujax/qujax_convert.py
Comment (cqc-melf) [inline pytket/extensions/qujax/qujax_convert.py:43]: What are this arguments? I think the name of the function is a bit unintuitive, I would expect additional argument that needs to be passed in, maybe the name could be changed to `_expended` or so, to make that more clear?

Comment (cqc-melf) [reply pytket/extensions/qujax/qujax_convert.py:43]: Have you thought about putting this all in the same function and changing the return value based on an additional parameter?

File: pytket/extensions/qujax/qujax_convert.py
Comment (SamDuffield) [reply pytket/extensions/qujax/qujax_convert.py:183]: I thought it would be useful to be able to print out a string representation of the circuit (i.e. when working in console not jupyter)

qujax does not require pytket so the pytket.Circuit -> print circuit functionality has to be hosted on the extension

File: pytket/extensions/qujax/qujax_convert.py
Comment (SamDuffield) [reply pytket/extensions/qujax/qujax_convert.py:43]: The "qujax_args" are (gate_seq, qubit_inds_seq, param_inds_seq, n_qubits) which define a circuit in qujax (there is no circuit class in qujax). I think the name is ok but I will make this clearer in the docstring.

Comment (SamDuffield) [reply pytket/extensions/qujax/qujax_convert.py:43]: Removing all of the *_symbolic functions and just having it inferred based on would be nice, and would clean up the code somewhat.

Currently we do not do this because the behaviour of `tk_to_qujax(circuit)` and `tk_to_qujax_symbolic(circuit, symbol_map=None)` is slightly different (in how it interprets parameterised gates) as described in the docstring of tk_to_qujax_symbolic.

As this change changes the API in a way that is not backwards compatible (it will break code that uses tk_to_qujax_symbolic), I suggest we leave this to a later release?

File: pytket/extensions/qujax/qujax_convert.py
Comment (cqc-melf) [reply pytket/extensions/qujax/qujax_convert.py:184]: I thought this might be better handled with a print of the converted circuit. So only offering the conversion in pytket-qujax and using the qujax print function? Not sure if this is a function that I would recommend to use, because the user can do something else with the circuit then printing the result of the conversion? The result of the conversion is hidden and if they want to do something else they would need to repeat the conversion?

File: pytket/extensions/qujax/qujax_convert.py
Comment (cqc-melf) [reply pytket/extensions/qujax/qujax_convert.py]: I would not be worried about breaking the API in this point, doing it now or later is both fine with me.

Comment (cqc-melf) [reply pytket/extensions/qujax/qujax_convert.py]: Have you thought about defining a qujax.circuit containing all this information and making this a bit more intuitive for the users? (No need to do that now, but might be good to do that some time in the future?)

File: pytket/extensions/qujax/qujax_convert.py
Comment (SamDuffield) [reply pytket/extensions/qujax/qujax_convert.py:184]: I think having both functionalities is ok, it is cumbersome to carry around the qujax_args and a key goal of this extension is to have the functionality of qujax indexed via a pytket.Circuit

My idea behind print_circuit is just a quick sanity check that the circuit is what you want it to be, on that basis I enjoy having the ability to do
`print_circuit(circuit)`
rather than
`qujax_args = tk_to_qujax_args(circuit)`
`print_circuit(*qujax_args)`

So really I see the main functions being `tk_to_qujax` and `print_circuit` with `tk_to_qujax_args` there for completeness and in case users do not want to repeat the conversion (although repeating the conversion isn't so bad as it's very fast)

File: pytket/extensions/qujax/qujax_convert.py
Comment (SamDuffield) [reply pytket/extensions/qujax/qujax_convert.py]: I would like to keep qujax without a qujax.circuit, JAX is built for functional programming (easier to debug, easier to compose with other code), so adding a custom class would go against this

So I think the workflow with the extension is nice for the user. Intuitive circuit manipulations can be done with pytket and then sent to qujax for fast implementation, differentiation, parallelisation etc

Comment (SamDuffield) [reply pytket/extensions/qujax/qujax_convert.py]: with regards to breaking the API, you are right probably the earlier the better

However, I know `tk_to_jax_symbolic` is currently used by some users so we should be wary

Comment (cqc-melf) [reply pytket/extensions/qujax/qujax_convert.py]: If they are currently using this, it should not be a blocker for us. They are not forced to upgrade?

File: pytket/extensions/qujax/qujax_convert.py
Comment (cqc-melf) [reply pytket/extensions/qujax/qujax_convert.py:184]: This question is what the sanity check is helping if you can't use what you have checked? If you want to do a check, there should not be the need to trust conversion function to give the same result every time?

Comment (cqc-melf) [reply pytket/extensions/qujax/qujax_convert.py:184]: Especially for symbolic circuits there could be a change depended in the parameters given?

Comment (SamDuffield) [reply pytket/extensions/qujax/qujax_convert.py:184]: the print_circuit function won't actually display the values of the parameters only the index of the input parameter array in which to find it for each gate

so the `print_circuit` is just a sanity check that the `pytket.Circuit` (and possibly `symbol_map`) have been constructed as intended. If something is amiss the `pytket.Circuit` can then be adjusted before being sent to `tk_to_qujax` which constructs the full JAX function

The user then doesn't ever have to touch any qujax_args (unless they want to)

Comment (cqc-melf) [reply pytket/extensions/qujax/qujax_convert.py:184]: Okay, thank you for making that clear!

File: pytket/extensions/qujax/qujax_convert.py
Comment (cqc-melf) [inline pytket/extensions/qujax/qujax_convert.py:273]:
```suggestion
    :type qubit_inds_seq: Sequence[Sequence[int]]
```

**Review by cqc-melf — CHANGES_REQUESTED** [review]

**Comment by cqc-melf** [comment]
Two more small issues in the docs, the rest looks good to me. Thank your for taking the time for the discussions!

**Comment by SamDuffield** [comment]
Thank you for your very helpful comments!

R.E. merging `tk_to_qujax_args_symbolic` into `tk_to_qujax_args`, this might be useful but currently the two functions work quite differently. I suggest we go ahead with this release (inc `tk_to_qujax_args_symbolic`) and I will raise the idea of removing `tk_to_qujax_args_symbolic` as a discussion in issues?

**Comment by cqc-melf** [comment]
Are we planning to do a release now after merging this? if not we should set the changelog title of 0.4.0 to `unrelease`

**Comment by SamDuffield** [comment]
> Are we planning to do a release now after merging this? if not we should set the changelog title of 0.4.0 to `unrelease`

Hmmmm I think there is a fair chance we decide to keep `tk_to_qujax_args_symbolic`, so I am pro new release and can then do a subsequent one if we do decide to consolidate?

**Review by cqc-melf — APPROVED** [review]

**Pull request merged**
